## Problem #164: Three Consecutive Digital Sum Limit
[Link to Problem](https://projecteuler.net/problem=164)

### Problem Description

How many $20$ digit numbers $n$ (without any leading zero) exist such that no three consecutive digits of $n$ have a sum greater than $9$?

### Approach

We can count the number of numbers that satisfy the property using **dynamic programming**. We have the $dp[len][a][b][c]$ = the number of numbers that satisfy the property with length $len$ and the last the digits are $ \overline{abc}$. The reccurence is easy to come up with.

In [8]:
L = 20

dp = [[[[0 for _ in range(10)] for _ in range(10)] for _ in range(10)] for _ in range(L + 1)]

for i in range(100, 1000):
    a, b, c = i // 100, i // 10 % 10, i % 10
    if a + b + c <= 9:
        dp[3][a][b][c] = 1

for len in range(4, L + 1):
    for a in range(10):
        for b in range(10):
            for c in range(10):
                for d in range(10):
                    if b + c + d <= 9:
                        dp[len][b][c][d] += dp[len - 1][a][b][c]

count = 0

for a in range(10):
    for b in range(10):
        for c in range(10):
            count += dp[L][a][b][c]

print(count)

378158756814587


###### Result: **378158756814587** | Execution time: 0s

### Complexity analysis

Time complexity: $O(L * 10^4)$

Space complexity: $O(L * 10^3)$

##### Tags: #dynamic-programming